### PR TITLE
Add subPath to PersistentClaimStorage.  Allows multiple users of same …

### PR DESCRIPTION
**Pull request by chonton**
Add subPath to PersistentClaimStorage.  Allows multiple users of same Volume, and workaround for ext4 lost+found interference

### Type of change
- Bugfix
- Enhancement / new feature
- Documentation

### Description

Kafka using the root directory of an ext4 volume has problem with the 'lost+found' directory.  Adding a subPath attribute to map Kafka usage.

### Checklist

- [ ] Update/write design documentation in `./design`
- [ ] Write tests
- [x] Make sure all tests pass
- [x] Update documentation
- [ ] Check RBAC rights for Kubernetes / OpenShift roles
- [ ] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [ ] Reference relevant issue(s) and close them after merging
- [ ] Update CHANGELOG.md

### CODE REVIEW

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/model/AbstractModel.java [hunk 1]
```diff
@@ -551,6 +551,16 @@ protected VolumeMount createVolumeMount(String name, String path) {
         return volumeMount;
     }
 
+    protected VolumeMount createVolumeMountWithSubPath(String name, String path, String subPath) {
```
Comment (scholzj) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/model/AbstractModel.java]: Do we really need two methods here? The old one without subPath and the new one with subPath? What does it mean when subPath is null? It calling `.withSubPath(null)` doesn't change anything, we can probably just use single method, or?

Comment (chonton) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/model/AbstractModel.java]: The two parameter method is used in many places.  I've updated the implementation of the two parameter method to invoke the three parameter method with a null subPath.

**Comment by chonton** [comment]
I've refactored implementation so that ZooKeeperCluster and KafkaCluster share the storage code.
The documentation has also been updated.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/model/AbstractModel.java [hunk 2]
```diff
@@ -734,64 +725,6 @@ protected Service createHeadlessService(List<ServicePort> ports, Map<String, Str
         return service;
     }
 
-    protected StatefulSet createStatefulSet(
```
Comment (chonton) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/model/AbstractModel.java]: Pushed down to StatefulCluster.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaCluster.java [hunk 3]
```diff
@@ -171,13 +164,6 @@
      */
     private Map<String, CertAndKey> brokerCerts;
 
-    /**
```
Comment (chonton) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaCluster.java]: Pushed down to StatefulCluster.

File: cluster-operator/src/test/java/io/strimzi/operator/cluster/model/KafkaClusterTest.java [hunk 4]
```diff
@@ -343,10 +343,14 @@ public void testPvcNames() {
...
-            assertEquals(kc.VOLUME_NAME + "-" + KafkaCluster.kafkaPodName(cluster, i),
-                    pvcs.get(0).getMetadata().getName() + "-" + KafkaCluster.kafkaPodName(cluster, i));
+        List<VolumeMount> vms = kc.getVolumeMounts();
+        for (int i = 0; i < pvcs.size(); ++i) {
+            String name = pvcs.get(i).getMetadata().getName();
+            assertEquals(kc.VOLUME_NAME, name);
+            VolumeMount vm = vms.get(i);
+            assertEquals(name, vm.getName());
+            assertEquals("/var/lib/kafka/" + name, vm.getMountPath());
+            assertNull(vm.getSubPath());
```
Comment (scholzj) [inline cluster-operator/src/test/java/io/strimzi/operator/cluster/model/KafkaClusterTest.java]: That is a valid test. But we should probably also have some test with the `subPath` set. I think it should be also either tested in `ZookeeperClusterTests` or done one level above in `StatefulClusterTests`.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/model/StatefulCluster.java [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+
+        StatefulSet statefulSet = new StatefulSetBuilder()
+                .withNewMetadata()
+                .withName(name)
+                .withLabels(getLabelsWithName(templateStatefulSetLabels))
+                .withNamespace(namespace)
+                .withAnnotations(mergeAnnotations(annotations, templateStatefulSetAnnotations))
+                .withOwnerReferences(createOwnerReference())
+                .endMetadata()
+                .withNewSpec()
```
Comment (scholzj) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/model/StatefulCluster.java]: Could you please keep the indentations which makes it a lot easier to read?

**Comment by chonton** [comment]
Updated with missing documentation.
Added unit tests and fixed two defects discovered with the unit tests.

**Comment by chonton** [comment]
This is not required with #1321.  Do you want to close?

**Comment by scholzj** [comment]
@chonton If you still see some value in this, I think it is fine to have this in Strimzi. While I personally see it as fairly limited in its usefulness, it might be useful for others. The #1321 was needed not only for the `lost+found` but also to restore the backwards compatibility. 

PS: Sorry I didn't got to review the latest changes - I hope to find some time for it soon.

**Comment by chonton** [comment]
Closing due to lack of time/interest.

**Pull request closed without merging**